keps/sig-storage/0661-statefulset-volume-resize/README.md

```diff
@@ -0,0 +1,685 @@
+motivation section can optionally provide links to [experience reports][] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+Allow users to resize the volumes associated with `StatefulSet` by modifying the size of 
+PVC in the `volumeClaimTemplate`. 
+
+<!--
+List the specific goals of the KEP.  What is it trying to achieve?  How will we
+know that this has succeeded?
+-->
+
+### Non-Goals
+
+Currently, PVC resize does not support shrinking the volume. Also, some environments require Pod restarts inorder 
+to get access to the resized volume. This proposal does not try to address these scenarios but relies on the 
+underlying PVC resize logic to take the appropriate action or indicate error as needed.
+
+<!--
+What is out of scope for this KEP?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+### Background
+
+Before we proceed to the proposal it is imperative to layout the various components involved 
+in creating and managing the persistent volumes associated with `StatefulSet`. StatefulSet 
+controller creates the PVCs from `volumeClaimTemplates` and associates them with the pods 
+spawned from it. Reconciliation logic runs for every change made to the StatefulSet 
+`Spec` and `Status` reflects the progress towards making the real world state equal to the
+ `Spec`. The admission controller in the api server is responsible for restricting the fields 
+which can be edited in a StatefulSet specification. The RBAC specification in the controller
+policy is responsible for adding the required RBAC permission for StatefulSet. StatefulSet 
+controller also supports revision control. 
+
+### Changes required
+
+The following changes are proposed to achieve the goal:
+1. Make the `volumeClaimTemplates` storage size editable by modification to the api server 
+   validation.
+2. Add reconciliation of the associated PVC's size and the individual `volumeClaimTemplates` 
+   size into the StatefulSet controller reconciliation. 
+3. Append PVC `update` to the StatefulSet controller's RBAC clusterrole.
+
+Once the above changes are made, modification to any of the `volumeClaimTemplates` storage 
+size by user will update the underlying PVC via the reconciliation performed by (2) above. 
+The existing PVC [resize](https://kubernetes.io/blog/2018/07/12/resizing-persistent-volumes-using-kubernetes/) 
+code kicks in at this point and attempts at the resize of the volume. 
+
+### Error indications to user
+
+Whether a resize works or not would depend on many factors such as is `allowExpansion` set in the 
+storage class, does the underlying storage allow expansion etc. Such failures will be reported in 
+the corresponding PVC events. StatefulSet would only report any failures in the update operation 
+performed to the PVC object.
+
+### Revision control
```

Comment:
So this has some implications for rollback and failure detection. This needs more thought. If we modify the volume size and the controller wedges midway in the rollout do to a failure, what would the remediation strategy be? There is no roll back to a previous revision. If the user modifies the spec of the StatefulSet to have the smaller volume size to attempt a rollback, that will not be possible either (i.e. no shrinking).

Reply:
Not true shrinking but a mechanism to allow user to cancel a previously issued volume expansion is in works - https://github.com/kubernetes/enhancements/issues/1790  . This requires users to reduce PVC size to previous capacity if they want to cancel the expansion.